### PR TITLE
 clippy: Map to an error type instead of using allowing `result_unit_err` in `components/url`

### PR DESCRIPTION
**Pull request by ektuu**
---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes are related to #31832 
- [x] These changes do not require tests because they do not change functionality.

### CODE REVIEW

**Review by Loirooriol — CHANGES_REQUESTED** [review]
Some points:
 - When you receive feedback in a PR, just update the old one, don't create a new one
 - The only difference from cd28798d7e72b4ccab4808dd111050545bcdf339 is that you avoided inserting a space before `#![deny(unsafe_code)]`, but you are still adding unnecessary newlines
 - Is there any reason to reorder the methods?
 - The template for pull requests has a list of things that you need to test, and check if they hold.
 - In particular, `./mach test-tidy` is complaining: https://github.com/servo/servo/actions/runs/8395608656/job/22995237911?pr=31834
 - What are the warnings that you are trying to address?

**Comment by ektuu** [comment]
Thankyou for your feedback. I will address the issues mentioned.

In the above PR, I used a custom error type to address the warnings of functions returning 'Result<_, ()>'. 
Methods such as set_username, set_ip_host, set_password, to_file_path, and from_file_path have been modified to return a Result type with UrlError instead of ().  Previously existing method definitions have been modified or removed to align with the new error handling strategy.

**Comment by ektuu** [comment]
@Loirooriol 
I've addressed the issues. can you please review?

**Review by Loirooriol — COMMENTED** [review]
What's the reason for reordering the methods?

**Comment by ektuu** [comment]
Methods have been reordered to align with the new error handling strategy. The methods which use custom error UrlError have been put together to enhance readability of the code. Would you want me to reorder them ?

File: components/url/lib.rs
Comment (eerii) [inline components/url/lib.rs:37]: A small nit, I think it would make sense if you used the already imported `fmt::` namespace for `Display`, `Formatter` and `Result`, as most of the files in the project already use this convention. If you decide to do it, it would also be nice to update the return type of `debug_compact` to match.

```suggestion
impl fmt::Display for UrlError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
```

On a more global note, I wonder if it is necessary to add this error type. This is wrapping `url`'s library functions, and they decided not to include an error type for [these functions](https://docs.rs/url/latest/url/struct.Url.html#method.set_username). As I saw it the backtrace in case of an error would be enough to determine which function was causing the issue, that's why I added the `allow`, but a case could be made that the error provides better context.

File: components/url/lib.rs
Comment (mrobinson) [inline components/url/lib.rs:46]: I think we can just remove this entirely as the `Debug` implementation should be enough.

File: components/url/lib.rs
Comment (mrobinson) [inline components/url/lib.rs:159]: Let's keep the methods in the same order. I'm not sure it makes sense to organize them by their return value.

**Review by mrobinson — CHANGES_REQUESTED** [review]

File: components/url/lib.rs
Comment (eerii) [reply components/url/lib.rs:46]: I think `Display` is there because `std::error::Error` [requires it](https://doc.rust-lang.org/std/error/trait.Error.html), but that can probably be removed as well. Other errors in servo such as `WebGLError` or `TimeRangesError` don't implement `Error` or `Display`, just `Debug`.

Comment (mrobinson) [reply components/url/lib.rs:46]: Makes sense. Let's remove both of these.

**Review by mrobinson — APPROVED** [review]

**Comment by mrobinson** [comment]
@ektuu Once a change has been added to the merge queue making any change to the branch will remove it from the merge queue. By clicking the "Update branch" button, you took the change off the merge queue. In general it is *not necessary* to click this button. I recommend rebasing your work locally instead of clicking this button.

**Comment by mrobinson** [comment]
Looks like I was wrong here. This didn't go to the merge queue, because it is still waiting on a review from @Loirooriol, as he requested changes earlier.

**Review by Loirooriol — APPROVED** [review]
I share eerii's doubts about whether this is actually an improvement, but I'm fine with the change.

**Comment by mrobinson** [comment]
I also agree that as it is, this isn't super useful, but it does prepare the code for better error handling in the future.

**Pull request merged**
